pkg/sabakan-state-setter/controller.go

```diff
@@ -199,27 +203,116 @@ func (c *Controller) runOnce(ctx context.Context) error {
 		log.Warn("failed to get sabakan machines", map[string]interface{}{
 			log.FnError: err.Error(),
 		})
-		// lint:ignore nilerr  RunPeriodically tries this again.
+		// lint:ignore nilerr  Run tries this again.
 		return nil
 	}
 	if len(machines) == 0 {
 		log.Warn("no machines found", nil)
 		return nil
 	}
 
+	// Check machine types
+	undefinedMachineTypes := map[string][]string{}
+	for _, m := range machines {
+		if _, ok := c.machineTypes[m.Type]; !ok {
+			undefinedMachineTypes[m.Type] = append(undefinedMachineTypes[m.Type], m.Serial)
+		}
+	}
+	for undefinedType, invalidMachines := range undefinedMachineTypes {
+		if undefinedType == "" {
+			log.Warn("machine type is not specified", map[string]interface{}{
+				"machines": strings.Join(invalidMachines, ","),
+			})
+		} else {
+			log.Warn("specified machine type does not exist", map[string]interface{}{
+				"machine-type": undefinedType,
+				"machines":     strings.Join(invalidMachines, ","),
+			})
+		}
+	}
+
+	// Do machines health check
+	newStateMap := c.machineHealthCheck(ctx, machines)
+
+	// Do machines retire
+	// T.B.D.
+
+	now := time.Now()
+	for _, m := range machines {
+		newState, ok := newStateMap[m.Serial]
+		if !ok || m.State == newState {
```

Comment:
The check of `m.State == newState` is new in this PR.
It seems reasonable, but I'm afraid there is some reason not to do this check in the old code...

By the way, `newState == m.State` would be preferable.

Reply:
I am one of the members who wrote the initial sabakan-state-setter.
As far as I remember, there is not any reason in the old code...

Reply:
Without the `!ok` check, the state of the StateUpdating, StateRetiring, or StateRetired machines will be updated unnecessarily.
So this condition needs `!ok`.

Reply:
Checking `!ok` is OK:-)
I have a feeling of strangeness in the order of the equality check, though it might be a matter of taste.

```suggestion
		if !ok || newState == m.State {
```